File: leetcode/Tree/leetcode-103/103.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }

        Queue<TreeNode> order = new LinkedList<>();
        Queue<TreeNode> reverse = new LinkedList<>();
        Stack<TreeNode> stack = new Stack<>();

        order.offer(root);

        while (!order.isEmpty() || !reverse.isEmpty()) {
            List<Integer> tmp1 = new ArrayList<>();
            while (!order.isEmpty()) {
                TreeNode node = order.poll();
                tmp1.add(node.val);
                if (node.left != null) {
                    reverse.offer(node.left);
                }
                if (node.right != null) {
                    reverse.offer(node.right);
                }
            }
            if (tmp1.size() > 0){
                res.add(tmp1);
            }
            
            List<Integer> tmp2 = new ArrayList<>();
            while (!reverse.isEmpty()) {
                TreeNode node = reverse.poll();
                stack.push(node);
                if (node.left != null) {
                    order.offer(node.left);
                }
                if (node.right != null) {
                    order.offer(node.right);
                }
            }
            while (!stack.isEmpty()) {
                tmp2.add(stack.pop().val);
            }
            if (tmp2.size() > 0) {
                res.add(tmp2);
            }
        }
        return res;
    }
}

File: leetcode/Tree/leetcode-103/103.go
package tree

import "container/list"

// TreeNode Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	res := [][]int{}
	if root == nil {
		return res
	}
	orderQ := list.New()
	reverseQ := list.New()

	orderQ.PushBack(root)

	for orderQ.Len() > 0 || reverseQ.Len() > 0 {
		tmp1 := []int{}
		for orderQ.Len() > 0 {
			node := orderQ.Front().Value.(*TreeNode)
			orderQ.Remove(orderQ.Front())
			tmp1 = append(tmp1, node.Val)
			if node.Left != nil {
				reverseQ.PushBack(node.Left)
			}
			if node.Right != nil {
				reverseQ.PushBack(node.Right)
			}
		}
		if len(tmp1) > 0 {
			res = append(res, tmp1)
		}

		tmp2 := []int{}
		for reverseQ.Len() > 0 {
			node := reverseQ.Back().Value.(*TreeNode)
			reverseQ.Remove(reverseQ.Back())
			tmp2 = append(tmp2, node.Val)
			if node.Right != nil {
				orderQ.PushFront(node.Right)
			}
			if node.Left != nil {
				orderQ.PushFront(node.Left)
			}
		}
		if len(tmp2) > 0 {
			res = append(res, tmp2)
		}
	}
	return res
}
